matrix/matrix.py

```diff
@@ -0,0 +1,202 @@
+               line_data):
+            raise ValueError("Matrix lines must be tuple")
+
+        if not all([len(line) == len(data) for line in data]):
+            raise ValueError("Matrix must be even-sided")
+
+    @classmethod
+    def unity(cls, dimension: int):
+        """Generates a matrix with a main diagonal of 1's, everything else
+        is 0's.
+        """
+        return cls(*[tuple((0,) * j + (1,) + (0,) * i) for j, i in enumerate(
+            range(dimension - 1, -1, -1))])
+
+    @classmethod
+    def ones(cls, dimension):
+        """Generates a matrix full of 1's."""
+        return cls(*([tuple((1,) * dimension) for _ in range(dimension)]))
+
+    def __repr__(self):
+        return super(Matrix, self).__repr__()
+
+    def __str__(self):
+        """Returns a string presentation of the matrix instance."""
+        return repr(self._matrix)
+
+    @property
+    def tuples(self):
+        """Returns the matrix instance as a tuple of tuples."""
+        return self._matrix
+
+    def _by_matrix(self, operation, dimension_of_other,
+                   self_item, other_item):
+        """Calculates operation on two items.
+
+        Args:
+            operation (function): the function of the operator.
+            dimension_of_other (int): the number of lines in the other matrix.
+            self_item (number): current item from the current matrix instance.
+            other_item (number): current item from the other matrix instance.
+
+        Returns:
+            number, the result of the operation.
+
+        Note:
+            Boolean operations supports different dimensions matrices but
+            mathematical operations don't (except matrix division).
+        """
+
+        if dimension_of_other != len(self.tuples) and not any(
+                [operation == operator.eq, operation == operator.ne]):
+            raise ValueError("Matrices length don't match")
+
+        if operation == operator.truediv:
+            raise ValueError(f"{operation} not supported on "
+                             f"{self.__class__.__name__}")
+
+        return operation(self_item, other_item)
+
+    def _by_scalar(self, operation, item, scalar):
+        """Calculates operation on item by scalar.
+
+        Args:
+            operation (function): the function of the operator.
+            item (number): current item from the current matrix instance.
+            scalar (number): the scalar to operate by.
+
+        Returns:
+            number, the result of the operation.
+
+        Note:
+            Scalar must be a number. Operations with scalars are only * and /.
+        """
+        if is_number(scalar) and not any([operation == operator.mul,
+                                          operation == operator.truediv]):
+            raise ValueError(f'{operation} not supported on {type(scalar)}')
+
+        return operation(item, scalar)
+
+    def _calculate_matrix_operation(self, other, operation):
```

Comment:
usually when writing code there is a trade-off
Generic code vs readable code.
Creating one function to rule them all (function that does everything), might be good as it support any action, but makes the code much less readable.

You can see that in your code, by the fact that everytime you call it, you must run validation on all operations, but the validation is not always valid for all operations, so you must add another if for it, etc.
This makes your code look clumsy